ledger/src/blockstore.rs

```diff
@@ -872,12 +896,59 @@ impl Blockstore {
         Ok(())
     }
 
+    fn write_shred(&self, shred: &Shred) {
+        let slot = shred.slot();
+        let path = if shred.is_data() {
+            self.data_shred_path(slot, shred.index() as u64)
+        } else {
+            self.coding_shred_path(slot, shred.index() as u64)
+        };
+        let mut tried = false;
+        loop {
+            info!(
+                "PUT_DATA: {}-{} {} {}",
+                slot,
+                shred.index(),
+                shred.is_data(),
+                solana_sdk::hash::hashv(&[&shred.payload]),
+            );
+            let e = Self::write_tmp(&path, &shred.payload);
```

Comment:
What happens if you get two duplicate shreds? the writes to the file may race

Reply:
There is a check up the pipeline to prevent dups.

Reply:
hmm, but I see `write_shred` is called directly from `insert_shreds` here: https://github.com/solana-labs/solana/blob/2c3efa307d0d80873ffec2b71c423e8da5a6c3b3/ledger/src/blockstore.rs#L949, where's the duplicate check?